bayes_opt1/util.py

```diff
@@ -1,10 +1,233 @@
+                diff[count] = diff[count] + (mean_n1star-mean_nstar)
+        count = count + 1
+    diff = diff/J
+    return diff
+
+
+# Implementation algorithm 4 (Frazier)
+
+def _kg4(x, optimizer, gp, n_grid = 100, J = 300):
+    '''
+    Library "numdifftools" needed:
+    
+                                    pip install numdifftools
+    REMARK: if p == 1, then x needs to be a np.float
+            if p > 1, then x needs to be a np.ndarray with of the form [,]
+    '''
+    dim = optimizer._space.bounds.shape[0]
+    
+    if dim == 1:
+        x = np.array(x)
+
+    if type(x) is not np.ndarray:
+        sys.exit('Error in _kg4: if p>1, x needs to be a NP.ndarray')
+    if len(x.shape) > 1 and x.shape[0] > 1:
+        sys.exit('Error in _kg4: x needs to have 1 row and p columns.')
+    
+    
+    
+    def mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,i):
+        '''
+        see Algorithm 4:
+        In our setting, we apply this by letting x_star_hat be the x^prime maximizing µ_{n+1}(x^prime; x, µ_n(x)+σ_n(x)Z), 
+        and then calculating the gradient of µ_{n+1}(x_star_hat; x, µ_n(x)+σ_n(x)Z) with respect to x 
+        while holding x_star_hat fixed.
+        '''
+        y=np.random.normal(loc=mean, scale=std)
+        x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+        y_newobs=np.append(y_obs,y)
+        mu=posterior(gp, x_newobs, y_newobs, x_star_hat)
+        return np.asscalar(mu)
+    
+    dim = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3) 
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, dim)
+
+    grid = grid_construction(optimizer._space.bounds, n_grid)
+    
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    grad_temp = []
+    mean, std= gp.predict(x.reshape(1,-1), return_std=True)
+    for j in range(J):
+        print("j = ",j+1)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,x)
+            mean_n1star = temp[0]
+            x_star_hat = grid[temp[1]].reshape(1,-1)
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(dim)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
+
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, pbounds, optimizer, gp):
+
+    dim = pbounds.shape[0]
+    x_T = []
+    
+    for r in range(R):
+        x_t = our_lhs(1, pbounds)[0]
```

Comment:
Attenzione, se chiamate `our_lhs()` dentro il loop chiedendole ogni volta un punto, significa che state ottenendo sempre lo stesso punto! Quello che voi volete è in realtà chiamarla fuori dal loop con `R` punti e nel loop prendere ogni volta l'`r`-esimo